modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,327 @@
+// function that takes the Maybe object and actually performs the operation.
+// This is done to make the functions easier to use with Fun.pipe, so that code
+// can be written left to right.
+//
+// Comparison with null or undefined
+// - We don't get fancy null coalescing operators with Maybe
+// - We do get fancy helper functions with Maybe
+// - Maybes support nesting, and allow for the type to still be nullable (or
+// another Maybe)
+// - There is no option to turn off strict-maybe-checks like there is for
+// strict-null-checks
+// - Try to use Maybe instead of null or undefined where you can
+//
+// Comparison with Optional
+// - Optional uses a.method().method(), Maybe uses Fun.pipe(a, method(), method())
+// - Optional is invariant, Maybe is covariant (Maybe<Dog> counts as a
+// Maybe<Animal> but Optional<Dog> does not count as Optional<Animal>)
+// - Maybe tree-shakes and minifies a lot better
+// - Maybe can be `console.log`ed
+// - Maybe might perform better because it doesn't need to create messy objects
+// - Optional is deprecated (but still in very wide use) and you should try to
+// migrate over to Maybe where you can
+
+// --- Types ---
+
+// Technically, this will compile down to 0 for Nothing, and 1 for Just. If
+// you're logging a `Maybe<T>` and you see some 1s and 0s where you expected to
+// see useful strings, this is why. This should minify better.
+export const enum Tag {
+  Nothing,
+  Just
+}
+
+export interface Nothing {
+  readonly tag: Tag.Nothing;
+}
+
+export interface Just<T> {
+  readonly tag: Tag.Just;
+  readonly value: T;
+}
+
+// Not exported, instead duplicated in Maybe.ts as an identical type to make
+// imports nice for consumers. Redefined here to avoid looped imports.
+type Maybe<T> = Nothing | Just<T>;
+
+// --- Identities ---
+
+// Just a small optimisation - all nothing objects are identical, why not make
+// them point to the same object?
+const singletonNothing: Nothing = { tag: Tag.Nothing };
+
+// Create a new Maybe<T> that **does not** contain a value. T can be any type
+// (and defaults to the literal any type) because we don't actually have a T.
+export const nothing = <T = any>(): Maybe<T> => singletonNothing;
+
+// Creates a new Maybe<T> that **does** contain a value.
+export const just = <T>(value: T): Maybe<T> => ({
+  tag: Tag.Just,
+  value
+});
+
+// Perform a transform on a Maybe type. Similar to the fold method on a number
+// of ADTs in the TinyMCE codebase, you pass two functions into this - if `self`
+// **does not** contain a value then `onNothing` will be called (with no
+// argument) but if `self` **does** contain a value then `onJust` will be
+// called (with the value as its argument).
+//
+// For the FP enthusiasts in the room, this function:
+// 1. Could be used to implement all of the functions below (although it isn't)
+// 2. Forms a catamorphism
+export const fold = <T, U>(onNothing: () => U, onJust: (value: T) => U) => (self: Maybe<T>): U => {
```

Comment:
Having just tried fp-ts I can tell you that this will not infer the function types nicely if you curry this.

Reply:
Discussed on slack, I will hopefully get used to pipe...